internal/contour/annotations.go

```diff
@@ -76,6 +78,27 @@ func parseAnnotationUInt32(annotations map[string]string, annotation string) *ty
 	return &types.UInt32Value{Value: uint32(v)}
 }
 
+// parseUpstreamProtocols parses the annotations map for a contour.heptio.com/upstream-protocol.{protocol}
+// where 'protocol' identifies which protocol must be used in the upstream.
+// If the value is not present, or malformed, then an empty map is returned.
+func parseUpstreamProtocols(annotations map[string]string, annotation string, protocols ...string) map[string]string {
+	up := make(map[string]string)
+	for _, protocol := range protocols {
+		ports, ok := annotations[fmt.Sprintf("%s.%s", annotation, protocol)]
+		if !ok {
```

Comment:
nit: you don't need to check ok here, the zero value of ports will be "", and splitting "" on "," will result in a slice of zero entries.

Reply:
good catch.

Reply:
Done.